specs/unified_kernel_image.md

```diff
@@ -146,3 +146,10 @@ that tools staging or consuming UKIs have a common place to store and look for t
 
 The installed UKIs should have a filename `<version format specification>.efi`, i.e. the filename is left to
 implementers but must be valid for comparisons according to the [Version Format Specification](version_format_specification.md).
+
+UKI addons (cmdline PE addons, systemd-sysext extensions for initrd) built centrally by distributions and
+installed via package manager should instead be installed in `/usr/lib/linux/extra.d/` if they are global
+addons/extensions, i.e. they are applied to all UKI installed in the ESP, and
+`/usr/lib/linux/$UNAME/$UKI.efi.extra.d/`, where `$UNAME` is the output of `uname -r` of the kernel included
+in the UKI, and `$UKI` is the UKI name just as in `$BOOT/EFI/Linux/$UKI.efi` and its addons folder
+`$BOOT/EFI/Linux/$UKI.efi.extra.d/` (see [`$BOOT`](./boot_loader_specification.md#the-boot-partition-placeholder)).
```

Comment:
I have a hard time parsing this sentence. Maybe breaking this into two sentences could make this easier to understand?

So the intention is:
1. Install a global UKI addon to `/usr/lib/linux/extra.d/`.
2. Install a UKI addon for a specific UKI to `/usr/lib/linux/$UNAME/$UKI.efi.extra.d/` and `$BOOT/EFI/Linux/$UKI.efi.extra.d/`.

Is that correct? I'm sorry if it's not the multiple ands confused me a bit.

If it's correct how about something like this
> UKI addons (cmdline PE addons, systemd-sysext extensions for the initrd) built centrally by distributions and
installed via package manager should be installed into locations depending on whether they are global UKI addons, that should be applied to all UKIs installed in the ESP, or whether they should only be applied to a single specific UKI. 
>
> Global UKI addons should be installed into `/usr/lib/linux/extra.d/`. UKI-specific addons should be installed to `/usr/lib/linux/extra.d/` and `$BOOT/EFI/Linux/$UKI.efi.extra.d/`, where `$UNAME` is the output of `uname -r` of the kernel included in the UKI, `$UKI` is the UKI name just as in `$BOOT/EFI/Linux/$UKI.efi`, and `$BOOT` is as it is defined in the [Boot Loader Specification](./boot_loader_specification.md#the-boot-partition-placeholder).

Reply:
No, I didn't want to mean that. But you are right, if it confuses, we need to write it better :smile: 
The intention is:
* global addons stay in `/usr/lib/linux/extra.d/`
* local addons stay in`/usr/lib/linux/$UNAME/$UKI.efi.extra.d/`.

What is `$UNAME`?  the output of `uname -r` of the kernel included in the UKI that is being referred by the local addon.
What is `$UKI`? It's the UKI name referred by the local addon, the same used in `$BOOT/EFI/Linux/$UKI.efi` .
What is `$BOOT`? it is defined in the [Boot Loader Specification](https://github.com/uapi-group/specifications/pull/boot_loader_specification.md#the-boot-partition-placeholder).

Reply:
So it should be
```
UKI addons (cmdline PE addons, systemd-sysext extensions for the initrd) built centrally by distributions and installed via package manager should be installed into locations depending on whether they are global UKI addons, that should be applied to all UKIs installed in the ESP, or whether they should only be applied to a single specific UKI.

Global UKI addons should be installed into `/usr/lib/linux/extra.d/`. UKI-specific addons should be installed to `/usr/lib/linux/$UNAME/$UKI.efi.extra.d/`, where `$UNAME` is the output of `uname -r` of the kernel included in the UKI, `$UKI` is the UKI name just as in $BOOT/EFI/Linux/$UKI.efi, and $BOOT is defined in the [Boot Loader Specification](https://github.com/uapi-group/specifications/pull/boot_loader_specification.md#the-boot-partition-placeholder).
```

Reply:
Yes, I think that's good. Thank you!

There remains the question of the path. There was a longer discussion in https://github.com/uapi-group/specifications/pull/60 that then settled to `/usr/lib/modules`. At least for sysexts as addons, I think the idea in that discussion was to have them go alongside. My understanding is, that you are choosing `/usr/lib/linux`, since there is prior art in Ubuntu (if I remember correctly). I can see arguments in either direction, especially for global addons, but for the UKI specific ones that is something that should be properly hashed out so that everyone is happy. 

I can see the following possibilities
- go with `/usr/lib/linux/$UNAME/$UKI.efi.extra.d/`
- go with /usr/lib/modules/$UNAME/$UKI.efi.extra.d/`
- allow both as independent directories
- allow symlinks
---
Having written all that, I have two thoughts:

1. > $UKI is the UKI name just as in $BOOT/EFI/Linux/$UKI.efi, and $BOOT is defined in the [Boot Loader Specification](https://github.com/uapi-group/specifications/pull/boot_loader_specification.md#the-boot-partition-placeholder).

I wonder if this could be made easier. We leave the name of the UKI in its place under `/usr/lib/modules` completely open, so should the maybe refer to that one instead of the one under `$BOOT`? Otherwise it is not necessarily clear what belongs together just from looking at `/usr/`.

2. Maybe this should also mention that the filenames for the addons should be valid to be compared by the version format specification as well.

Reply:
having the UKI in one hierarchy and the addons in another doesn't make much sense to me, we should use the same location for both (I do not mind which one personally)

Reply:
This same thought occourred to me yesterday too. I commented in https://github.com/systemd/systemd/issues/29372#issuecomment-1983260633 as we are mainly discussing it there.
 
Regarding `$UKI`, I think it's not a huge problem what exactly that is pointing. In my opinion, a smart choice would be to have `$UKI` in `/usr/lib/modules/$UKI` and in `$BOOT` to be the same, because it will be easy for the user and the tools coming up to pair the UKIs.
If we want to somehow force it through documentation is another point though.

I don't get point 2.

Reply:
I think the "too" is unnecessary, because the extension is different.
> The installed UKI addons should have a filename `<version format specification>.addon.efi`, i.e. the filename is left to implementers but must be valid for comparisons according to the Version Format Specification.